File: bg_locale/page/translation_manager/translation_manager.js
frappe.pages["translation-manager"].on_page_load = function (wrapper) {
  const page = frappe.ui.make_app_page({
    parent: wrapper,
    title: "Translation Manager",
    single_column: true,
  });

  const state = {
    language: "bg",
    search: "",
    missing_only: 0,
    limit: 200,
    offset: 0,
    dirty: new Map(), // key: row.name or synthetic key
  };

  const $controls = $(`
    <div class="row" style="margin-bottom: 12px;">
      <div class="col-sm-2">
        <label>Language</label>
        <input class="form-control" type="text" value="bg" data-field="language" />
      </div>
      <div class="col-sm-6">
        <label>Search (source contains)</label>
        <input class="form-control" type="text" placeholder="e.g. Submit, Customer..." data-field="search" />
      </div>
      <div class="col-sm-2">
        <label>&nbsp;</label>
        <div class="checkbox">
          <label><input type="checkbox" data-field="missing_only" /> Missing only</label>
        </div>
      </div>
      <div class="col-sm-2" style="display:flex; gap:8px; align-items:flex-end;">
        <button class="btn btn-primary" data-action="reload">Reload</button>
        <button class="btn btn-success" data-action="save">Save</button>
      </div>
    </div>
  `);

  $(page.body).append($controls);

  const $tableWrap = $(`<div></div>`);
  $(page.body).append($tableWrap);

  let dt = null;

  function getControl(name) {
    return $controls.find(`[data-field="${name}"]`);
  }

  async function loadRows() {
    state.language = (getControl("language").val() || "bg").trim();
    state.search = (getControl("search").val() || "").trim();
    state.missing_only = getControl("missing_only").is(":checked") ? 1 : 0;

    const r = await frappe.call({
      method: "bg_locale.bg_locale.page.translation_manager.translation_manager.get_translations",
      args: {
        language: state.language,
        search: state.search,
        missing_only: state.missing_only,
        limit: state.limit,
        offset: state.offset,
      },
      freeze: true,
      freeze_message: "Loading translations...",
    });

    const rows = (r.message && r.message.rows) || [];

    renderTable(rows);
  }

  function renderTable(rows) {
    $tableWrap.empty();

    const data = rows.map((x) => ({
      name: x.name,
      source_text: x.source_text,
      translated_text: x.translated_text || "",
      context: x.context || "",
      modified: x.modified,
    }));

    const columns = [
      { name: "source_text", label: "Source Text", width: 420 },
      { name: "translated_text", label: "Translated Text (edit)", width: 420, editable: true },
      { name: "context", label: "Context", width: 160 },
      { name: "modified", label: "Modified", width: 160 },
    ];

    dt = new frappe.DataTable($tableWrap.get(0), {
      columns,
      data,
      layout: "fixed",
      inlineFilters: false,
      noDataMessage: "No translations found.",
      cellHeight: 42,
      dynamicRowHeight: true,
      events: {
        onCellEdit: function (cell, value) {
          const row = dt.datamanager.getRow(cell.rowIndex);
          // Track dirty changes by Translation docname
          state.dirty.set(row.name, {
            name: row.name,
            source_text: row.source_text,
            translated_text: row.translated_text,
            context: row.context,
          });
        },
      },
    });
  }

  async function saveDirty() {
    if (state.dirty.size === 0) {
      frappe.msgprint("Nothing to save.");
      return;
    }

    const items = Array.from(state.dirty.values());

    for (const item of items) {
      await frappe.call({
        method: "bg_locale.bg_locale.page.translation_manager.translation_manager.upsert_translation",
        args: {
          language: state.language,
          source_text: item.source_text,
          translated_text: item.translated_text,
          context: item.context || null,
        },
        freeze: true,
        freeze_message: "Saving...",
      });
    }

    state.dirty.clear();
    frappe.show_alert({ message: "Saved.", indicator: "green" });
    await loadRows();
  }

  $controls.on("click", '[data-action="reload"]', () => loadRows());
  $controls.on("click", '[data-action="save"]', () => saveDirty());

  // initial load
  loadRows();
};

File: bg_locale/page/translation_manager/translation_manager.py
import frappe
from frappe import _


def _is_allowed():
    # MVP: only System Manager
    return "System Manager" in frappe.get_roles(frappe.session.user)


@frappe.whitelist()
def get_translations(language="bg", search="", missing_only=0, limit=200, offset=0):
    if not _is_allowed():
        frappe.throw(_("Not permitted"), frappe.PermissionError)

    missing_only = int(missing_only or 0)
    limit = min(int(limit or 200), 500)
    offset = max(int(offset or 0), 0)

    filters = {"language": language}
    # Show only rows that exist in Translation table.
    # "Missing only" means translated_text empty.
    if missing_only:
        filters["translated_text"] = ["in", ["", None]]

    # search in source_text
    search = (search or "").strip()
    or_filters = None
    if search:
        # OR filters require query builder or get_all with "or_filters"
        or_filters = [{"source_text": ["like", f"%{search}%"]}]

    rows = frappe.get_all(
        "Translation",
        filters=filters,
        or_filters=or_filters,
        fields=["name", "language", "source_text", "translated_text", "context", "modified"],
        order_by="modified desc",
        limit_start=offset,
        limit_page_length=limit,
    )

    return {"rows": rows}


@frappe.whitelist()
def upsert_translation(language, source_text, translated_text, context=None):
    if not _is_allowed():
        frappe.throw(_("Not permitted"), frappe.PermissionError)

    language = (language or "").strip()
    source_text = (source_text or "").strip()
    translated_text = (translated_text or "").strip()
    context = (context or "").strip() or None

    if not language or not source_text:
        frappe.throw(_("Language and Source Text are required."))

    existing = frappe.get_all(
        "Translation",
        filters={"language": language, "source_text": source_text, "context": context},
        fields=["name"],
        limit=1,
    )

    if existing:
        doc = frappe.get_doc("Translation", existing[0]["name"])
        doc.translated_text = translated_text
        doc.save(ignore_permissions=True)
    else:
        doc = frappe.get_doc(
            {
                "doctype": "Translation",
                "language": language,
                "source_text": source_text,
                "translated_text": translated_text,
                "context": context,
            }
        )
        doc.insert(ignore_permissions=True)

    # Make changes show up immediately
    frappe.clear_cache()

    return {"ok": True, "name": doc.name}
